### PR TITLE
Add OAuth 2.0 Documentation

### PR DESCRIPTION
**Pull request by justin-shipstation**
This documentation has been needed for awhile, all of the examples were pulled from mono auth for production integrations. 
The main meat of the documentation is thanks to @bill-robbins-ss and his amazing documentation in the mono auth repository

### CODE REVIEW

File: src/pages/docs/oauth/index.mdx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  - ShipEngine Connect
+  - OAuth
+---
+
+# Mono Auth Overview
+The ShipEngine Connect platform uses an internal system called `MonoAuth` to handle Oauth 1 and 2 auth handshake flows.
+This internal system works with a DSL that you define inside of your application definition, and supports property injection in the form of tags.
+
+## Installation Process
+- Define your application as using OAuth as seen in the following [example]()
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/index.mdx:14]: missing link

File: src/pages/docs/oauth/index.mdx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- Publish your application to ShipEngine Connect using the `connect publish`
+- Contact a member of the [ShipEngine Connect Team]() in order to establish a MonoAuth installation
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/index.mdx:16]: missing link

File: src/pages/docs/oauth/index.mdx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    - Provide a ClientId & ClientSecret used for initiating the oauth workflow
+- Once the installation has been created you will recieve back a callback url Ex: `https://authbridge-sec.auctane.com/authflow/{application_code}/accept`
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/index.mdx:18]: does the user know how to obtain their `{application_code}` ?

File: src/pages/docs/oauth/flow/authorization.mdx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    "nonce": {
+        "name": "state"
+    }
+}
+```
+
+In the example above the Definition contains templated information describing
+how MonoAuth should construct and redirect the user's browser. 
+
+:brain: See the [OAuth 2.0 RFC - Authorization Request](https://tools.ietf.org/html/rfc6749#section-4.1.1)
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/flow/authorization.mdx:47]: :brain: doesn't render :)

File: src/pages/docs/oauth/flow/authorization.mdx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+for more information on what standards expect.
+
+- `url_template` The authorize endpoint of the integration
+- `query_parameters` Any query parameters that the integration's OAuth authorize
+endpoint expects to receive.
+- `nonce` A random string included in the authorization request and returned in the redirect. [This is used to prevent cross-site request forgery.](https://datatracker.ietf.org/doc/html/rfc6749#section-10.12)
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/flow/authorization.mdx:53]: add a snippet or link explaining `{system:authorize_callback_url}`

File: src/pages/docs/oauth/flow/request-token.mdx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+# Request Token
+
+After the end-user grants the Application access the integration redirects back
+to MonoAuth with a `code`. Monoauth makes a server-server request to the 3rd
+party to exchange the code for an access token.
+The `request_token` component of the DSL describes how MonoAuth should
+construct and execute that request.
+
+:brain: See the [OAuth 2.0 RFC - Access Token Request](https://tools.ietf.org/html/rfc6749#section-4.1.3)
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/flow/request-token.mdx:17]: 🧠 not working :)

File: src/pages/docs/oauth/flow/request-token.mdx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        "name": "client_id",
+        "value": "{installation:client_id}"
+      },
+      {
+        "name": "client_secret",
+        "value": "{installation:client_secret}"
+      },
+      {
+        "name": "code",
+        "value": "{callback:code}"
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/flow/request-token.mdx:45]: good opportunity to describe and link to the `{callback:}` expression documentation

**Comment by lzhang** [comment]
We need to take out any language referencing "monoauth", that's an internal codename of no relevance to the consumers of this dependency.

**Review by bill-robbins-ss — COMMENTED** [review]
> We need to take out any language referencing "monoauth", that's an internal codename of no relevance to the consumers of this dependency.

Was going to same the same thing, I think referring it to an Auth Bridge services may be acceptable, as the url subdomain for SEC is `authbridge-sec`

File: src/pages/docs/oauth/flow/refresh-token.mdx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+Some, but not all, 3rd parties return short-lived Access Tokens that must be
+refreshed using a "refresh token". The `refresh_token` component describes the
+server-server interaction MonoAuth can make to refresh an access token using a
+refresh token.
+
+The `refresh_token` flow differs from the other components of the auth flow in
+that it does not require any interaction in the user's browser. The MonoAuth
+client (Application) makes a server-to-server request to the *internal* endpoint
+`/internal/api/<integration-api-code>/<app-id>/refresh_token`.
+
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/flow/refresh-token.mdx:20]: Remove mention of the client and internal endpoints.

```suggestion
The `refresh_token` flow differs from the other components of the auth flow in
that it does not require any interaction in the user's browser. The ShipEngine Connect Platform will handle any token refreshing that is required.
```

File: src/pages/docs/oauth/flow/refresh-token.mdx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        "value": "{installation:client_id}"
+      },
+      {
+        "name": "client_secret",
+        "value": "{installation:client_secret}"
+      }
+    ]
+  }
+```
+
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/flow/refresh-token.mdx:54]: Good opportunity to explain/link to the expression `{result_token_response:`

File: src/pages/docs/oauth/flow/refresh-token.mdx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+<Pager
+  prev="./request-token.mdx" prevTitle="Request Token"
+  next="../index.mdx" nextTitle="Mono Auth Overview"
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/flow/refresh-token.mdx:57]: link to templating next

File: src/pages/docs/oauth/index.mdx [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+## Advanced Topics
+- [Response Transformation](./advanced/response-transformation.mdx) *(specify how responses from the two exchanges can be mapped into the MonoAuth canonical model)*
+- [Connection Context](./advanced/connection-context.mdx) *(allows a response transformer to specify a collection of properties in connection_context)*
+
+## Templating
+
+The Integration Definition DSL JSON is processed with a simple templating engine
+that supports variable substitution and some basic operations.
+
+The sample data in [dev_integration_data.json](../sample-data/dev_integration_data.json)
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/index.mdx:71]: dead link

File: src/pages/docs/oauth/resolver/base64.mdx [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+:brain: Note that the template engine itself is recursive so template expressions
+can be evaluated: 
+
+```json
+// dev_installation_data.json
+{
+    "app_Id": "sample-application",
+    "integration_api_code": "sample-server",
+    "client_id": "dummy-client-id",
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/resolver/base64.mdx:23]: remove this client-id and client-secret :)

**Comment by justin-shipstation** [comment]
> We need to take out any language referencing "monoauth", that's an internal codename of no relevance to the consumers of this dependency.

Ok @joshuaflanagan, @bill-robbins-ss , @lzhang is there a preferred term for describing mono auth to outside consumers?

File: src/pages/docs/oauth/resolver/installation.mdx [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+The `installation` Value Resolver references the Installation of an Integration
+within a MonoAuth (GTM) Application.
+
+```json
+// dev_installation_data.json
+{
+    "app_Id": "sample-application",
+    "integration_api_code": "sample-server",
+    "client_id": "dummy-client-id",
+    "client_secret": "dummy-client-secret"
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/resolver/installation.mdx:20]: remove client-id and client-secret values in this page

File: src/pages/docs/oauth/advanced/response-transformation.mdx [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+`IntegrationTokenResponse.TokenType` property with a value of `Bearer`.
+1. `expires_at` Is an object, rather than a JSONPath.
+    * The `path` property is JSONPath informing MonoAuth where to obtain the
+    value.
+    * The `date_time_format` property is a date-time format string that can be
+    either a [standard](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings) format string in .NET or a
+    [custom](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings) format string for .NET. This example
+    uses a custom format string.
+
+:warning: `expires_in` vs `expires_at` OAuth2 specifies `expires_in` as a standard
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/advanced/response-transformation.mdx:78]: ⚠️  is not rendering

File: src/pages/docs/oauth/advanced/connection-context.mdx [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+# Connection Context
+
+Nonstandard OAuth integrations occasionally both send and expect more than an
+access token in the response and authenticated requests, respectively.
+
+To accommodate for this, MonoAuth allows a response transformer to specify
+a collection of properties in `connection_context`. Properties defined in
+`connection_context` are included in the 'code for token' (`POST /authflow/token`)
+exchange response (the
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/advanced/connection-context.mdx:17]: give an example of a monoauth response here, as the link to the C# POCO won't work :)

File: src/pages/docs/oauth/advanced/connection-context.mdx [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+object).
+
+Values in the `connection_context` can use some _template expressions_, like
+[`{callback:`](./TEMPLATING_REF.md#callback). This allows
+an integration definition to capture values from the `redirect_uri` from the
+3rd party to include in the connection context.
+For more information on this feature see
+[PLAT-3246](https://auctane.atlassian.net/browse/PLAT-3246).
+
+The name `connection_context` comes from the ShipEngine Connect Platform and
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/advanced/connection-context.mdx:28]: get rid of iPaaS, unless SEC users know what it is

File: src/pages/docs/oauth/advanced/connection-context.mdx [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+iPaaS as a whole. `connection_context` may include data entered by the user
+from the data-driven connection UI described by an integration and/or data
+extracted by the `connection_context` response transformer in an integration's
+MonoAuth DSL.
+
+:warning: Including _any_ response transformer shifts the burden of response
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/advanced/connection-context.mdx:34]: ⚠️ not rendering :)

File: src/pages/docs/oauth/advanced/connection-context.mdx [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+Values in the `connection_context` can use some _template expressions_, like
+[`{callback:`](./TEMPLATING_REF.md#callback). This allows
+an integration definition to capture values from the `redirect_uri` from the
+3rd party to include in the connection context.
+For more information on this feature see
+[PLAT-3246](https://auctane.atlassian.net/browse/PLAT-3246).
+
+The name `connection_context` comes from the ShipEngine Connect Platform and
+iPaaS as a whole. `connection_context` may include data entered by the user
+from the data-driven connection UI described by an integration and/or data
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/advanced/connection-context.mdx:30]: link to Forms https://connect-git-update-oauth-shipengine.vercel.app/docs/v2/structure#forms and get rid of data-driven

**Review by bill-robbins-ss — COMMENTED** [review]
Great start ! 

Consider defining DSL for the users up front as the acronym is used a decent amount in the rest of the docs

**Comment by joshuaflanagan** [comment]
> > We need to take out any language referencing "monoauth", that's an internal codename of no relevance to the consumers of this dependency.
> 
> Ok @joshuaflanagan, @bill-robbins-ss , @lzhang is there a preferred term for describing mono auth to outside consumers?

As @bill-robbins-ss mentioned, the DNS uses the term "auth bridge" - but I'm not sure we really want to emphasize that either. Do we need a name for it? Can we get by with using more generic language like "the system".

For example:

> To accommodate for this, MonoAuth allows a response transformer to specify a collection of properties in...

could be

> To accommodate for this, the system allows a response transformer to specify a collection of properties in...

or "we allow". The point is, the consumer doesn't need to know there is a separate named service doing this work - its just "the system". Or even, "ShipEngine Connect".

**Comment by justin-shipstation** [comment]
> .

So the problem is we need to communicate what it is that DSL is driving, if not the documentation becomes a bit weird.

File: src/pages/docs/oauth/index.mdx (hunk 3 above)
Comment (justin-shipstation) [reply src/pages/docs/oauth/index.mdx:18]: Yes, I was letting them know that they will be given back a url shaped like that.

**Comment by justin-shipstation** [comment]
@joshuaflanagan @bill-robbins-ss I have gone ahead and removed all instances of IPaaS or mention of MonoAuth

File: src/pages/docs/oauth/index.mdx (hunk 3 above)
Comment (bill-robbins-ss) [reply src/pages/docs/oauth/index.mdx:18]: I don't think they even need to know about this URL.

File: src/pages/docs/oauth/flow/refresh-token.mdx [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+---
+
+# Refresh Token
+
+Some, but not all, 3rd parties return short-lived Access Tokens that must be
+refreshed using a "refresh token". The `refresh_token` component describes the
+server-server interaction Auth Bridge can make to refresh an access token using a
+refresh token.
+
+The `refresh_token` flow differs from the other components of the auth flow in
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/flow/refresh-token.mdx:16]: ```The `refresh_token` flow differs from the other components of the auth flow in
that it does not require any interaction in the user's browser. The Auth Bridge
client (Application) makes a server-to-server request to the *internal* endpoint
`/internal/api/<integration-api-code>/<app-id>/refresh_token`.```
remove this section

File: src/pages/docs/oauth/examples/trade-me.mdx [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+Trade Me uses a 3 legged OAuth 1 strategy, so for them we needed to make use of the auth1 redirector in `advanced_configuration`
+``` JSON
+"advanced_configuration": [
+    {
+      "name": "redirector",
+      "value": "oauth1"
+    }
+  ]
+```
+It also needed to take advantage of [response transformation](../advanced/response-transformation.mdx) to map the response payload from the `RequestToken` endpoint.
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/examples/trade-me.mdx:19]:
```suggestion
It also needs to take advantage of [response transformation](../advanced/response-transformation.mdx) to map the response payload from the `RequestToken` endpoint.
```

File: src/pages/docs/oauth/examples/trade-me.mdx [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+### Example Response Payload:
+``` JSON
```
Comment (bill-robbins-ss) [inline src/pages/docs/oauth/examples/trade-me.mdx:21]: this is missing `access_token_secret`

**Pull request merged**